packages/@aws-cdk/aws-servicecatalogappregistry/lib/application.ts

```diff
@@ -134,20 +141,26 @@ abstract class ApplicationBase extends cdk.Resource implements IApplication {
   /**
    * Associate stack with the application in the stack passed as parameter.
    *
-   * If the stack is already associated, it will ignore duplicate request.
    * A stack can only be associated with one application.
    */
   public associateApplicationWithStack(stack: cdk.Stack): void {
     if (!this.associatedResources.has(stack.node.addr)) {
-      new CfnResourceAssociation(stack, 'AppRegistryAssociation', {
+      const stackCondition = stack.nestedStackResource?.cfnOptions.condition;
+      const association = new CfnResourceAssociation(stack, 'AppRegistryAssociation', {
         application: stack === cdk.Stack.of(this) ? this.applicationId : this.applicationName ?? this.applicationId,
         resource: stack.stackId,
         resourceType: 'CFN_STACK',
       });
 
-      this.associatedResources.add(stack.node.addr);
-      if (stack !== cdk.Stack.of(this) && this.isSameAccount(stack) && !this.isStageScope(stack)) {
-        stack.addDependency(cdk.Stack.of(this));
+      if (stackCondition) {
+        association.addOverride('Condition', stackCondition.logicalId);
```

Comment:
What you have described is the exact situation here, AWS Solutions team had a stack with conditional nested stack. While they used our L2 construct, the association failed as the condition evaluated to be `false` whereas our L2 construct previously couldnt handle that condition in `ResourceAssociation`. This is a fix for the same. I will add this detail to the PR description.

Reply:
I understand that it has something to do with putting conditions on NestedStacks, but you missed one part of my question: if I recall the design is for the `ResourceAssocation` constructs to be INSIDE the stacks that are being associated.

* If that is the case, there is no need to associate the condition with the association.
* If that is NOT the case, I think you might be fixing the wrong bug (i.e., it would be a more correct fix to move the association to inside the stack).

Allow me to explain with a diagram. This is what I think should be going on, purely looking at the CloudFormation template level:

![image](https://user-images.githubusercontent.com/524162/207267549-db0daace-8090-4ae0-95b1-4ffdd2ef1ec9.png)

If that is not what is going on (if instead the `ResourceAssociation` is in `ParentStack` instead of `ChildStack`), then:

* We need to fix that; or
* We need an explanation for why the pattern is different for nested stacks.

Reply:
you are right, the diagram is correct and the problem never happens for the new `ApplicationAssociator` construct. However, this problem occurs only if customers try to explicitly associate a nested child stack like below:

```
const application = new appreg.Application(stack, 'AutoApplication', {
    applicationName: 'MyAutoApplication',
});
application.associateApplicationWithStack(parentStack);
application.associateApplicationWithStack(childStack); // this is where it fails as the stack itself has not been created and our service throws the exception.
```

Reply:
Actually you are right, since `associateApplicationWithStack` is creating the association inside child stack, we dont need to add the condition for `ResourceAssociation`.